### ISSUE TITLE
flow_from_dataframe implementation

### ISSUE THREAD

**Issue by matobler** [rstudio/keras#658]
It would be great if you could create an R function the flow_from_dataframe generator that has recently been added to Keras. We have several 100k images with associated labels in a text file that we would like to use for training. We could move them into class directories but during model development and testing labels get updated and sometimes we need multiple labels for each image (e.g. category and count).

**Comment by matobler** [comment]
`flow_images_from_dataframe<-function (dataframe, directory=NULL,x_col="filename",y_col="class",
                                      generator = image_data_generator(), target_size = c(256,256), 
                                      color_mode = "rgb", classes = NULL, class_mode = "categorical", 
                                      batch_size = 32, shuffle = TRUE, seed = NULL, save_to_dir = NULL, 
                                      save_prefix = "", save_format = "png",dtype="float32",
                                      subset = NULL, interpolation = "nearest",drop_duplicates=TRUE) 
{
  args <- list(dataframe=as.data.frame(dataframe),directory = normalize_path(directory),
               x_col=x_col,y_col=y_col,
               target_size = as.integer(target_size), 
               color_mode = color_mode, classes = classes, class_mode = class_mode, 
               batch_size = as.integer(batch_size), shuffle = shuffle, 
               seed = as_nullable_integer(seed), save_to_dir = normalize_path(save_to_dir), 
               save_prefix = save_prefix, save_format = save_format,dtype=dtype, 
               drop_duplicates=drop_duplicates)
  if (keras_version() >= "2.1.2") 
    args$interpolation <- interpolation
  if (keras_version() >= "2.1.5") 
    args$subset <- subset
  do.call(generator$flow_from_dataframe, args)
}
environment(flow_images_from_dataframe) <- asNamespace('keras')`

I managed to figure out how to write the function. Trained some models with it already and it works great. Just make sure you have the pandas module installed.

Please feel free to include this function into the keras package.

**Comment by andrewsris** [comment]
Hello, thank you for your help! I am just learning Keras, and I chose R over python mainly because I am familiar with working with dataframes. Thus the function that you defined will definitely help me in creating multiple models to test on the same set of images.

However, I am still very new to python and keras, and I am having difficulty getting this function to work. In Rstudio, how would I check to see if I have the pandas module installed? When using the function, I get an error that says:

Error in py_get_attr_impl(x, name, silent) : AttributeError: 'ImageDataGenerator' object has no attribute 'flow_from_dataframe'

**Comment by matobler** [comment]
It is possible that you have an older version of Keras installed. I use version 2.2.4 and keras-preprocessing version 1.0.9. If you have pip you can check the version with 
```
pip show keras
pip show keras-preprocessing
```
Update keras and keras-preprocessing and it should work.

For pandas type 
`pip show pandas`

you can install it with
`pip install pandas`

If you have a stand alone installation of Python but don't have pip follow this: 
https://www.liquidweb.com/kb/install-pip-windows/
If you use Anaconda these commands are different.

**Comment by andrewsris** [comment]
Thank you very much. I do have pandas 0.24.2 installed. I re-ran install_keras() and it shows that I now have installed keras-2.2.4, keras-applications-1.0.7, keras-preprocessing-1.0.9. I do think that was a large part of the issue, as it seems like I was likely running an old version of keras. I ran your function again, and the error is different. Now the error says:

Error in py_call_impl(callable, dots$args, dots$keywords) : 
  AttributeError: 'list' object has no attribute 'apply'

I feel this is likely an error on my end regarding my syntax for using the function. Do you possibly have an example of how to use your function? Thanks!

**Comment by andrewsris** [comment]
Well for what it's worth, I figured out my issue and it looks like your flow_images_from_dataframe function is now working well for me. The issue I was having wasn't anything with your code. As it turns out, I didn't realize that my RStudio client wasn't loading in the virtual environment that I had set up for tensorflow and python3.6. It was loading python 2.7 which I do not believe was compatible with your function. By setting the RETICULATE_PYTHON path to point towards python 3, I was able to get this working. 

(Furthermore I found out that the reticulate package's use_python() function does not work very well in RStudio. Setting the RETICULATE_PYTHON prior to loading the reticulate and keras packages worked)
